### PR TITLE
Comment Template Block: Add test coverage for context setting and inner block insertion

### PR DESCRIPTION
**Pull request by ockham**
Backport of https://github.com/WordPress/gutenberg/pull/50879 and https://github.com/WordPress/gutenberg/pull/50883.

Trac ticket: https://core.trac.wordpress.org/ticket/58839

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by ockham** [comment]
Note that the `test_rendering_comment_template_sets_comment_id_context` test will only pass once a version of `@wordpress/block-library` is used in Core that includes https://github.com/WordPress/gutenberg/pull/50279.

**Comment by ockham** [comment]
Added tests from https://github.com/WordPress/gutenberg/pull/50883.

**Comment by tellthemachines** [comment]
Hi 👋  
I assume this change is meant for 6.3? If so, could you create a Trac ticket to link this PR please?

**Comment by ockham** [comment]
@tellthemachines Apologies, I still haven't gotten around to do this. Upon second thought, I think I should split the PR in two, one of which we should be able to merge before Beta 1 (when blocks are synced from GB), whereas the other test will only pass after that sync.

I'll try to take care of that splitting, plus filing the required tickets, early next week. (Then again, none of this is super-urgent, so it's not the end of the world if it doesn't go into 6.3.)

**Comment by tellthemachines** [comment]
Thanks for the update @ockham !

**Comment by ockham** [comment]
> Note that the `test_rendering_comment_template_sets_comment_id_context` test will only pass once a version of `@wordpress/block-library` is used in Core that includes https://github.com/WordPress/gutenberg/pull/50279.

Rebased to pick up the updated packages and to make tests pass 🤞 

Wondering if we should consider including these with 6.3 after all, given https://core.trac.wordpress.org/ticket/58699. (The tests wouldn't have caught that, but they provide some confidence that we don't introduce any regressions when fixing that. This might be a bit more academic here, since the [fix](https://github.com/WordPress/gutenberg/pull/52364) has to be made in GB anyway, where those tests are already in place.)

**Comment by tellthemachines** [comment]
I'm not against adding more test coverage 😄 

Is this ready for review then? Asking because it's still in draft mode.

**Comment by ockham** [comment]
> I'm not against adding more test coverage 😄
> 
> Is this ready for review then? Asking because it's still in draft mode.

Ah yep, I guess it is -- apologies 😅 I guess I wasn't 100% sure if the tests were "minimal" enough (especially `test_inner_block_inserted_by_render_block_data_is_retained`), but maybe that's for reviewers to decide 😅

File: tests/phpunit/tests/blocks/renderCommentTemplate.php
Comment (andrewserong) [inline tests/phpunit/tests/blocks/renderCommentTemplate.php]: This refers to a `gutenberg` filter it looks like, should this be removed?

File: tests/phpunit/tests/blocks/renderCommentTemplate.php
Comment (andrewserong) [inline tests/phpunit/tests/blocks/renderCommentTemplate.php]: As above, does this need to be here?

**Review by andrewserong — COMMENTED** [review]
These tests are generally looking good to add to me. Just left a question about the `remove_filter` and `add_filter` calls for Gutenberg filters. I imagine those can be removed? I couldn't find `gutenberg_auto_insert_child_block` in the Gutenberg repo, but I might be missing some context there potentially.

File: tests/phpunit/tests/blocks/renderCommentTemplate.php
Comment (andrewserong) [reply tests/phpunit/tests/blocks/renderCommentTemplate.php]: Ah, looks like these might be leftovers from an experiment over in https://github.com/WordPress/gutenberg/pull/51294 ? The tests pass for me fine locally with these lines removed for this PR.

File: tests/phpunit/tests/blocks/renderCommentTemplate.php
Comment (ramonjd) [inline tests/phpunit/tests/blocks/renderCommentTemplate.php:524]: `@ticket` and `@covers` where appropriate (I know the trac ticket is TBD! 😄 )

File: tests/phpunit/tests/blocks/renderCommentTemplate.php
Comment (peterwilsoncc) [inline tests/phpunit/tests/blocks/renderCommentTemplate.php]: These three assertions will need messages to help differentiate between failures.

File: tests/phpunit/tests/blocks/renderCommentTemplate.php
Comment (peterwilsoncc) [inline tests/phpunit/tests/blocks/renderCommentTemplate.php]: CS Nit

```suggestion
			/*
			 * Insert a Comment Author Name block (which requires `commentId`
			 * block context to work) after the Comment Content block.
			 */
```

**Comment by ockham** [comment]
> These tests are generally looking good to add to me. Just left a question about the `remove_filter` and `add_filter` calls for Gutenberg filters. I imagine those can be removed? I couldn't find `gutenberg_auto_insert_child_block` in the Gutenberg repo, but I might be missing some context there potentially.

Ah, great spot! Yep, they can be removed; I'll take care of that. Thank you for flagging 👍 😄

File: tests/phpunit/tests/blocks/renderCommentTemplate.php
Comment (ockham) [reply tests/phpunit/tests/blocks/renderCommentTemplate.php]: https://github.com/WordPress/wordpress-develop/pull/4507/commits/c6257da39005ea91363287b4507669b7432c6688

File: tests/phpunit/tests/blocks/renderCommentTemplate.php
Comment (ockham) [reply tests/phpunit/tests/blocks/renderCommentTemplate.php:516]: https://github.com/WordPress/wordpress-develop/pull/4507/commits/c6257da39005ea91363287b4507669b7432c6688

File: tests/phpunit/tests/blocks/renderCommentTemplate.php
Comment (ockham) [reply tests/phpunit/tests/blocks/renderCommentTemplate.php]: https://github.com/WordPress/wordpress-develop/pull/4507/commits/c6257da39005ea91363287b4507669b7432c6688

File: tests/phpunit/tests/blocks/renderCommentTemplate.php
Comment (ockham) [reply tests/phpunit/tests/blocks/renderCommentTemplate.php]: a4add40810

File: tests/phpunit/tests/blocks/renderCommentTemplate.php
Comment (costdev) [inline tests/phpunit/tests/blocks/renderCommentTemplate.php]: No need to store the callback, just pass it directly here. The test suite cleans up all filters after each test, so no need to call `remove_filter()`

File: tests/phpunit/tests/blocks/renderCommentTemplate.php
Comment (ockham) [reply tests/phpunit/tests/blocks/renderCommentTemplate.php:524]: 3ad0fb7bd8

File: tests/phpunit/tests/blocks/renderCommentTemplate.php
Comment (ockham) [reply tests/phpunit/tests/blocks/renderCommentTemplate.php]: Oh, you mean like inline? Cool, will do 👍

Comment (ockham) [reply tests/phpunit/tests/blocks/renderCommentTemplate.php]: 756caa65ab

**Comment by ockham** [comment]
Took me a while, but I’ve filed the [ticket](https://core.trac.wordpress.org/ticket/58839). I've also addressed everyone's feedback (thank you!)

We're pretty close to RC1, so I'm not sure if folks are comfortable merging this so late in the game. OTOH, it's a test-only change that's supposed to guard against regressions 🤔

**Comment by ockham** [comment]
Added some more coverage in d3e3ffd880 to cover the scenario fixed by https://github.com/WordPress/gutenberg/pull/52364.

cc/ @dlh01

File: tests/phpunit/tests/blocks/renderCommentTemplate.php
Comment (mukeshpanchal27) [inline tests/phpunit/tests/blocks/renderCommentTemplate.php]:
```suggestion
	 * @ticket 58839
	 *
	 * @covers ::render_block_core_comment_template
```

File: tests/phpunit/tests/blocks/renderCommentTemplate.php
Comment (mukeshpanchal27) [inline tests/phpunit/tests/blocks/renderCommentTemplate.php]:
```suggestion
	 * @ticket 58839
	 *
	 * @covers ::render_block_core_comment_template
```

File: tests/phpunit/tests/blocks/renderCommentTemplate.php
Comment (costdev) [inline tests/phpunit/tests/blocks/renderCommentTemplate.php]:
```suggestion
	 * Tests that the Comment Template block makes comment ID context available to programmatically inserted child blocks.
```

File: tests/phpunit/tests/blocks/renderCommentTemplate.php
Comment (costdev) [inline tests/phpunit/tests/blocks/renderCommentTemplate.php]:
```suggestion
	 * Tests that an inner block added via the render_block_data filter is retained at render_block stage.
```

**Review by costdev — APPROVED** [review]
Just some nit feedback, but with these and @mukeshpanchal27's suggestions, this looks good to go. Thanks @ockham!

Pre-approving so as not to hold up commit as my internet is tempramental today.

**Review by felixarntz — APPROVED** [review]

**Review by mukeshpanchal27 — APPROVED** [review]

**Comment by ockham** [comment]
Thanks all! Committed to Core in https://core.trac.wordpress.org/changeset/56262 😊

**Pull request closed without merging**
